### DISCUSSION TITLE
Does the CLI cache results?

### DISCUSSION THREAD

**Discussion by adriangb** [microsoft/pyright#4809] [Q&A]
I'm trying to run pyright (CLI, installed via npm) on a relatively large codebase. It takes ~50s the first run, which is actually great compared to mypy. But subsequent runs take the same amount of time even if there are no changes to the source files. Does pyright have a cache, and if so how can I debug why it is not using it or why it takes the same amount of time to run again with no changes?

**Comment by erictraut** [comment]
Pyright maintains an in-memory cache. If you run the CLI with the `--watch` option, it will watch for file changes and reanalyze the portions of the project that are affected by any changes. This should be much faster than analyzing from scratch most of the time. Pyright does not maintain any on-disk cache. In my experience, such caches are very problematic. (Refer to the many cache coherency issues that plague mypy, for example.) I also find that persistent caches are unnecessary if you make the analyzer fast enough. For those reasons, we have no plans to add any persistent caches to pyright.

**Reply by adriangb** [reply]
Thank you, that explains what I'm observing. We're trying to get pyright to replace mypy as a pre-commit hook, so I'm not sure the `--watch` option would work. I know that pyright has a server component, would it be possible to start the server in the background and then communicate with it from the pre-commit hook so that the errors are displayed as if they had been executed from the pre-commit hook?

**Reply by erictraut** [reply]
Pyright's internal architecture does include a server component, but it's not exposed externally in the CLI. It's used by the CLI's watch mode, which uses file server watchers to determine when a file has changed. It's also used by the LSP (language server protocol) implementation, which receives messages from a language client when a file is modified. I suppose you could fork pyright and create a version of it that runs as an independent process. It could accept external commands that indicate which files have been modified, but that's pretty far afield of how pyright was intended to work, and you'd need to maintain it yourself.

My team has never used pre-commit hooks. Instead, we use either the watch mode in the CLI or a language server client (like VS Code), which supports `"diagnosticMode": "workspace"` to continually perform type checking for all files in the workspace. Using either of these techniques, you are guaranteed to have visibility into any errors detected by pyright prior to attempting to check in a change.

**Reply by adriangb** [reply]
Thanks for the info. While I do think it would be interesting to get pyright better supported as a pre-commit hook I do understand that would be difficult and certainly do not want to support it myself. We're open to other ways of working.

> Using either of these techniques, you are guaranteed to have visibility into any errors detected by pyright prior to attempting to check in a change.

What do you mean by visibility? I tried enabling `"diagnosticMode": "workspace"` and I don't get any errors when trying to check in a change, which makes sense because as far as I can tell nothing is hooking into the commit process. Other than checking the `Problems` pane in VSCode manually before commiting, what does the visibility look like? The nice thing about a pre-commit hook is that there is no manual process, it just stops you from commiting.

**Reply by erictraut** [reply]
> What do you mean by visibility?

I mean that the programmer will see any errors reported prior to attempting to check in a change. The errors will be visible during the editing process, either in a shell window (if running the CLI in watch mode) or in the editor's Problems tab (if running the language server). 

Perhaps you're used to a different workflow, but I personally always wanted to be made aware of any errors during the "inner loop" of development — while I'm editing. If there are still errors in my code at the point where I want to check in a fix, then something has gone very wrong. If I somehow fail to notice that there are errors and proceed with a commit (something that I admit happens on very rare occasions), then my cloud-based CI scripts catch the problem.

**Reply by adriangb** [reply]
Gotcha, thanks. I think we just have slightly different development processes. I do monitor errors while editing, but also like to have type checkers (or the compiler in other languages) run when I try to check in my changes before I have to wait 5 minutes for CI to report back. For what it's worth I think this workflow is relatively typical, at least amongst the open-source projects I work on.

**Reply by erictraut** [reply]
Is your code base public, by any chance? If so, I could see if there's any low-hanging fruit for reducing the 50-second analysis time. Or you could see if there are any files that stand out as being especially expensive by running `pyright --stats --verbose`.

**Reply by adriangb** [reply]
That would be amazing. The codebase I'm talking about is pydantic's codebase. We have two primary repos:
- https://github.com/pydantic/pydantic-core
- https://github.com/pydantic/pydantic

`pydantic-core` is particularly interesting from a type checker perspective, we have some large recursive unions: https://github.com/pydantic/pydantic-core/blob/main/pydantic_core/core_schema.py. MyPy gets pretty tripped up by these, pyright is quite fast on just that module. Where pyright takes 50 seconds is if you try to run it on the `/pydantic` folder in the `pydantic` repo. I set up [this branch](https://github.com/pydantic/pydantic/tree/pyright) to run pyright in CI, here's a sample run: https://github.com/pydantic/pydantic/actions/runs/4473072857/jobs/7859999089?pr=5232. The errors are not important, we could fix those, the issue is the 1 min + runtime, specifically when used locally as a commit guard (1-2 minutes in CI would be no problem).

The main areas that I think slow things down is the extensive use of those recursive tagged unions in [this module](https://github.com/pydantic/pydantic/blob/pyright/pydantic/json_schema.py), [this module](https://github.com/pydantic/pydantic/blob/pyright/pydantic/_internal/_generate_schema.py) and other places around the codebase.

I attached the output from a run with `--verbose`
[output.txt](https://github.com/microsoft/pyright/files/11023266/output.txt)


Let me know if there's anything else I can do to make it easier to dig around the codebase.

**Reply by erictraut** [reply]
Ah, what a coincidence. I spent some time this past week looking at [type checking performance for pydantic](https://github.com/microsoft/pyright/issues/4781). Pyright version 1.1.300, which I'll publish tomorrow, will incorporate these improvements ([this one](https://github.com/microsoft/pyright/commit/99c73a0f7da617e93c609087ee36841be11bcdde) and [this one](https://github.com/microsoft/pyright/commit/99c73a0f7da617e93c609087ee36841be11bcdde)). On my machine, these two changes reduced the overall pydantic type checking time from 17s to 4s. The main optimizations were around the use of large unions. In the case of pydantic, it uses a 40-element union called `CoreSchema`, and pyright had some O(n^2) algorithms related to such unions. My change makes these O(n) in most cases.

I'm running on a Mac with a relatively fast M2 processor, so the times you see will likely still be bigger than the 4s I'm observing, but I'm hopeful that you'll see times around 10s once you update to pyright 1.1.300.

**Reply by adriangb** [reply]
Amazing work! Looks like one of the links should be [this one](https://github.com/microsoft/pyright/commit/c7ab8045bae66a19b69ce013de6f8fb642ec403c) (they're both the same right now). Just curious what prompted you to look into pydantic's source code?

Does pyright try to infer that this is a tagged union (via the `type` key in this case) and use that to it's advantage in any way? It seems like MyPy definitely does not, they have what appears to be worse than O(n^2) runtime.

As an aside, we've been having discussions with the pydantic team about speeding up data validation for unions. I imagine what we do or are thinking of doing internally is similar to what type checkers have to do. One of the things we've been looking into is speeding up non-tagged unions by analyzing simple keys that eliminate a lot of the union options (like a `type: Literal[...]` key or existence of keys that only appear in some of the union members) before doing more expensive checks (like checking a `users: List[SomeComplexUserSchema]` which might require checking every item in the list). In some sense this would be like doing a BFS instead of a DFS. Presumably one could even analyze the schemas beforehand to determine what the optimal comparison paths are. One the other hand, there's something to be said for having a simple algorithm along the lines of "check each union member in order, validating fields in the order they are defined" which would allow developers to order their fields and unions to optimize the path taken.

**Reply by erictraut** [reply]
I was prompted to look at this because of [this mypy issue](https://github.com/python/mypy/issues/14034).

> Does pyright try to infer that this is a tagged union

No, nothing like that. The main optimization is where a union type is being assigned to another union type. I optimized the cases where the two unions are the same or the "receiving" type is a union that contains all of the same elements as the "source" type plus some additional elements. This case turns out to be quite common across code bases, but it doesn't matter that much until the number of elements in the union grows to a large number.

**Reply by adriangb** [reply]
Might be of interest: https://github.com/pydantic/pydantic/commit/d491968f711038051eecd48d651835288a8c98bc. This was in great part because of those speedups you implemented.

**Reply by travisdowns** [reply]
For what it's worth, we are also looking at integrating `pyright` into our pre-commit hook workflow and I agree with @adriangb comments above that this a common workload.

We _also_ use the LSP in our IDEs to detect these errors in the "inner loop", but there are cases where you wouldn't notice it in the editor, e.g., when edits in one file cause errors in another you haven't even opened. Probably that's what `"diagnosticMode": "workspace"` is for (you would notice errors across files) but currently our codebase is checked at varying levels of strictness (which we are not able to encode in a single `pyrightconfig.json`) so we use multiple runs of pyright to check at different levels of strictnes.

Some files take 20+ seconds to type-check alone, making them infeasible to run in pre-commit.
